bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/FileRecoveryLog.java

```diff
@@ -0,0 +1,237 @@
+     * @param recoveryLogDir The directory to store the recovery log files in.
+     * @param deleteOldLogs Whether to delete old log files when creating a new one.
+     */
+    public FileRecoveryLog(String baseFileName, int checkpointInterval, Path recoveryLogDir, boolean deleteOldLogs) {
+        this.baseFileName = baseFileName;
+        this.recoveryLogDir = recoveryLogDir;
+        this.deleteOldLogs = deleteOldLogs;
+        this.checkpointInterval = checkpointInterval;
+        this.existingLogs = new HashMap<>();
+        this.file = createNextVersion();
+        this.numOfPutsSinceLastCheckpoint = 0;
+    }
+
+    /**
+     * Creates the next version of the recovery log file, cleaning up the previous one.
+     *
+     * @return The newly created log file.
+     */
+    private File createNextVersion() {
+        int latestVersion = findLatestVersion();
+        File oldFile = recoveryLogDir.resolve(baseFileName + latestVersion + ".log").toFile();
+        if (oldFile.exists()) {
+            existingLogs = readLogsFromFile(oldFile);
+            if (deleteOldLogs) {
+                File[] files = recoveryLogDir.toFile().listFiles(
+                        (dir, name) -> name.matches(baseFileName + "(\\d+)\\.log")
+                );
+                for (File file: files) {
+                        file.delete();
+                    }
+                }
+            }
+        File newFile = recoveryLogDir.resolve(baseFileName + (latestVersion + 1) + ".log").toFile();
+        try {
+            Files.createDirectories(recoveryLogDir); // create directory if not exists
+            newFile.createNewFile();
+            initAppendChannel(newFile);
+            if (existingLogs == null) {
+                return newFile;
+            }
+            // write existing unfinished logs to the new file
+            if (!existingLogs.isEmpty()) {
+                cleanUpFinishedLogs();
+                if (!existingLogs.isEmpty()) {
+                    for (Map.Entry<String, TransactionLogRecord> entry : existingLogs.entrySet()) {
+                        put(entry.getValue());
+                    }
+                    existingLogs.clear();
+                }
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to create recovery log file in " + recoveryLogDir);
+        }
+        return newFile;
+    }
+
+    /**
+     * Finds the latest version of the recovery log file.
+     *
+     * @return The latest version of the recovery log file.
+     */
+    private int findLatestVersion() {
+        int latestVersion = 0;
+        File directory = recoveryLogDir.toFile();
+        File[] files = directory.listFiles((dir, name) -> name.matches(baseFileName + "(\\d+)\\.log"));
+        if (files == null) {
+            return latestVersion;
+        }
+        for (File file : files) {
+            String fileName = file.getName();
+            int version = Integer.parseInt(fileName.replaceAll(
```

Comment:
Lets make sure that even a file name that doesn't adhere to the expected format will not cause any unexpected error here.